crates/blockchain-tree/src/shareable.rs

```diff
@@ -95,6 +95,21 @@ impl<DB: Database, C: Consensus, EF: ExecutorFactory> BlockchainTreeViewer
         self.tree.read().block_indices().canonical_chain().clone()
     }
 
+    fn find_canonical_ancestor(&self, hash: BlockHash) -> Option<BlockHash> {
+        let mut parent = hash;
+        let tree = self.tree.read();
+
+        // walk up the tree and check if the parent is canonical
+        loop {
+            if let Some(block) = tree.block_by_hash(parent) {
+                parent = block.parent_hash;
+            }
+            if tree.block_indices().is_block_hash_canonical(&parent) {
+                return Some(parent)
+            }
+        }
```

Comment:
what if the parent hash is not known to the tree? seems like the loop would never terminate

Reply:
you're right, forgot the break

Reply:
ah yeah was going to comment this